semanticdb/semanticdb3/javacp.md

```diff
@@ -0,0 +1,480 @@
+
+<table>
+  <tr>
+    <td width="220px"><b>Category</b></td>
+    <td><b>Examples</b></td>
+  </tr>
+  <tr>
+    <td valign="top">Primitive types [<a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.2">75</a>]</td>
+    <td>
+      <ul>
+        <li><code>byte</code> ~ <code>TypeRef(None, &lt;scala.Byte#&gt;, List())</code>.</li>
+        <li><code>short</code> ~ <code>TypeRef(None, &lt;scala.Short#&gt;, List())</code>.</li>
+        <li><code>int</code> ~ <code>TypeRef(None, &lt;scala.Int#&gt;, List())</code>.</li>
+        <li><code>long</code> ~ <code>TypeRef(None, &lt;scala.Long#&gt;, List())</code>.</li>
+        <li><code>char</code> ~ <code>TypeRef(None, &lt;scala.Char#&gt;, List())</code>.</li>
+        <li><code>float</code> ~ <code>TypeRef(None, &lt;scala.Float#&gt;, List())</code>.</li>
+        <li><code>double</code> ~ <code>TypeRef(None, &lt;scala.Double#&gt;, List())</code>.</li>
+        <li><code>boolean</code> ~ <code>TypeRef(None, &lt;scala.Boolean#&gt;, List())</code>.</li>
+        <li><code>void</code> ~ <code>TypeRef(None, &lt;scala.Unit#&gt;, List())</code>.</li>
+      </ul>
+    </td> </tr>
+  <tr>
+    <td valign="top">Type variable [<a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.4">78</a>]</td>
+    <td>
+      <ul>
+        <li><code>T</code> ~ <code>TypeType(List(), None, None)</code>.</li>
+        <li><code>T extends S</code> ~ <code>TypeType(List(), None, Some(&lt;S&gt;))</code>.</li>
+        <li><code>T extends S1 &amp; ... &amp; Sn </code> ~ <code>TypeType(List(), None, Some(IntersectionType(&lt;S1&gt;, ..., &lt;Sn&gt;)))</code>.</li>
+      </ul>
+    </td>
+  </tr>
+  <tr>
+    <td valign="top">Parameterized types [<a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.5">79</a>]</td>
+    <td>
+      <ul>
+        <li><code>C&lt;T1, ..., Tn&gt;</code> ~ <code>TypeRef(None, &lt;C&gt;, List(&lt;T1&gt;, ..., &lt;Tn&gt;))</code>.</li>
+        <li><code>C&lt;?&gt;</code> ~ <code>TypeRef(None, &lt;C&gt;, List(&lt;local_wildcard&gt;))</code>.</li>
+        <li><code>C&lt;? extends T&gt;</code> ~ <code>TypeRef(None, &lt;C&gt;, List(&lt;local_wildcard&gt;))</code>.</li>
+      </ul>
+    </td>
+  </tr>
+  <tr>
+    <td valign="top">Array types [<a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html#jls-10.1">80</a>]</td>
+    <td>
+      <ul>
+        <li><code>T[]</code> ~ <code>TypeRef(None, &lt;scala.Array#&gt;, List(&lt;T&gt;))</code>.</li>
+        <li><code>int[]</code> ~ <code>TypeRef(None, &lt;scala.Array#&gt;, List(&lt;scala.Int#&gt;))</code>.</li>
+      </ul>
+    </td>
+  </tr>
+  <tr>
+    <td valign="top">Intersection types [<a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.9">81</a>]</td>
+    <td>
+      <ul>
+        <li><code>T1 &amp; ... &amp; Tn</code> ~ <code>IntersectionType(&lt;T1&gt;, ..., &lt;Tn&gt;)</code>.</li>
+      </ul>
+    </td>
+  </tr>
+</table>
```

Comment:
Is it intentional that not all subsections of ch4 are covered here?

Reply:
1) Intersection types are part of the type variable `<tr>` now, but in JLS they are separate, so I think they deserve a separate `<tr>` in our spec.

2) Raw types are not an internal typechecker construct. They can be used as is in source code.

3) I understand that reference types and raw types are kinda covered in the parameterized types `<tr>`, but I believe that it would be beneficial to have a dedicated `<tr>` for every relevant section of the spec to make sure that the relationship between JLS and the SemanticDB spec is crystal clear. This is what I did in the Scala part of our spec.

Reply:
1. I removed the intersection example in the Type Variable section. Intersection types are already covered in the Intersection Types row.
2. Added a row for "Raw types" with a simple `T ~ TypeRef(None, <T>, List())` example
3. Added reference type row with basic examples for class/type variable/array types.

I agree it's good to have a 1:1 mapping with JLS, even if it means redundant sections.